File: Sen/Add_Two_Numbers.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        c1, c2 = l1, l2
        sentinel = ListNode(0)
        d = sentinel
        sum = 0

        while c1 or c2: # no matter if l1 and l2 have the same length
        	sum /= 10 # integer division, if last pass's sum is equal to or greater than 10, this pass's sum add 1
        	if c1:
        		sum += c1.val
        		c1 = c1.next 
        	if c2:
        		sum += c2.val
        		c2 = c2.next
        	d.next = ListNode(sum % 10)
        	d = d.next

        if sum / 10 == 1: # if the last pass's sum is equal to or greater than 10, create a new node with value 1
        	d.next = ListNode(1)

        return sentinel.next

File: ming/0920/Add Two Numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode newHead = new ListNode(0);
        ListNode curr = newHead;
        int carryOn = 0;
        while (l1 != null && l2 != null){
            int sum = l1.val + l2.val + carryOn;
            curr.next = new ListNode(sum % 10);
            carryOn = sum / 10;
            curr = curr.next;
            l1 = l1.next;
            l2 = l2.next;
        }
        while (l1 != null){
            curr.next = new ListNode((l1.val + carryOn) % 10);
            carryOn = (l1.val + carryOn) / 10;
            curr = curr.next;
            l1 = l1.next;
        }
        while (l2 != null){
            curr.next = new ListNode((l2.val + carryOn) % 10);
            carryOn = (l2.val + carryOn) / 10;
            curr = curr.next;
            l2 = l2.next;
        }
        if (carryOn == 1){
            curr.next = new ListNode(1);
        }
        return newHead.next;
    }
}